### PR TITLE
[Enhancement] Get tablet schema by index id in lake::DeltaWriter

### PR DESCRIPTION
**Pull request by sduzh**
- Get tablet schema by index id
- More comments
- More tests

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by kevincai — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 71 / 90 (78.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/async_delta_writer.h|0|3|00.00%|[163, 164, 165]|
|:large_blue_circle:|src/storage/storage_engine.cpp|0|1|00.00%|[1395]|
|:large_blue_circle:|src/storage/lake/delta_writer.h|0|3|00.00%|[167, 168, 169]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|15|22|68.18%|[615, 618, 672, 673, 675, 676, 677]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|54|59|91.53%|[264, 269, 270, 276, 318]|
|:large_blue_circle:|src/storage/lake/tablet.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 73 / 92 (79.35%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/async_delta_writer.h|0|3|00.00%|[163, 164, 165]|
|:large_blue_circle:|src/storage/storage_engine.cpp|0|1|00.00%|[1401]|
|:large_blue_circle:|src/storage/lake/delta_writer.h|0|3|00.00%|[167, 168, 169]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|15|22|68.18%|[615, 618, 672, 673, 675, 676, 677]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|56|61|91.80%|[267, 272, 273, 279, 337]|
|:large_blue_circle:|src/storage/lake/tablet.cpp|2|2|100.00%|[]|

**Review by gengjun-git — DISMISSED** [review]

File: be/src/storage/lake/tablet_manager.cpp [hunk 1]
```diff
@@ -670,6 +657,29 @@ StatusOr<TabletSchemaPtr> TabletManager::get_tablet_schema(int64_t tablet_id, in
...
+    // else: Cache miss, read the schema file
+    auto schema_file_path = join_path(tablet_root_location(tablet_id), schema_filename(index_id));
+    auto schema_or = load_and_parse_schema_file(schema_file_path);
+    TEST_SYNC_POINT_CALLBACK("get_tablet_schema_by_index_id.2", &schema_or);
+    if (schema_or.ok()) {
+        VLOG(3) << "Got tablet schema of id " << index_id << " for tablet " << tablet_id;
+        schema = std::move(schema_or).value();
+        // Save the schema into the in-memory cache, use the schema id as the cache key
+        auto cache_value = std::make_unique<CacheValue>(schema);
+        fill_metacache(global_cache_key, cache_value.release(), 0);
```
Comment (wyb) [inline be/src/storage/lake/tablet_manager.cpp:676]: why the cache value size is 0?

Comment (sduzh) [reply be/src/storage/lake/tablet_manager.cpp:676]: The schema object is managed by the GlobalTabletSchemaMap

Comment (wyb) [reply be/src/storage/lake/tablet_manager.cpp:676]: line 656 and 957
auto cache_size = `inserted ? (int)schema->mem_usage() : 0;`

0 if inserted else schema->mem_usage()?

Comment (sduzh) [reply be/src/storage/lake/tablet_manager.cpp:676]: How about modifying it in a separate PR? I just realized the logic, as well as the logic you mentioned in line 656 and line 957, may both be improper.

Comment (wyb) [reply be/src/storage/lake/tablet_manager.cpp:676]: ok

**Review by wyb — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 73 / 92 (79.35%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/async_delta_writer.h|0|3|00.00%|[163, 164, 165]|
|:large_blue_circle:|src/storage/storage_engine.cpp|0|1|00.00%|[1401]|
|:large_blue_circle:|src/storage/lake/delta_writer.h|0|3|00.00%|[167, 168, 169]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|15|22|68.18%|[615, 618, 672, 673, 675, 676, 677]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|56|61|91.80%|[268, 273, 274, 280, 338]|
|:large_blue_circle:|src/storage/lake/tablet.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 73 / 92 (79.35%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/async_delta_writer.h|0|3|00.00%|[163, 164, 165]|
|:large_blue_circle:|src/storage/storage_engine.cpp|0|1|00.00%|[1401]|
|:large_blue_circle:|src/storage/lake/delta_writer.h|0|3|00.00%|[167, 168, 169]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|15|22|68.18%|[615, 618, 672, 673, 675, 676, 677]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|56|61|91.80%|[268, 273, 274, 280, 338]|
|:large_blue_circle:|src/storage/lake/tablet.cpp|2|2|100.00%|[]|

**Pull request merged**